### PR TITLE
Backport recoilToTop to CMSSW 10_6_35

### PR DESCRIPTION
**Pull request by vslokenb**
#### PR description:

Backport of recoilToTop (recently pushed to CMSSW_13_2_X). Pushed from 10_6_35.

#### PR validation:

Same as linked original PR below. Verified using `scram build code-check` and `scram build code-format`. Seem to be a couple of minor indentation changes in lines unrelated to the actual TopRecoilHook in Pythia8Hadronizer.cc as a result of the code-format tool.  All differences between original PR are to account for CMSSW_10_6_35 formatting.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Original PR found [here](https://github.com/cms-sw/cmssw/pull/42180). Backport meant to allow for legacy Run 2 analysis, as this feature causes a shift in some top mass reconstruction variables.

### CODE REVIEW

**Comment by perrotta** [comment]
backport of #42180

**Comment by menglu21** [comment]
@cmsbuild please test

**Comment by menglu21** [comment]
+1

**Comment by rappoccio** [comment]
hold

Please prepare back ports of this to 13_0, 13_1, and 13_2.

**Comment by perrotta** [comment]
@vslokenb please also update the code according to https://github.com/cms-sw/cmssw/pull/42313

**Comment by menglu21** [comment]
please test

**Comment by menglu21** [comment]
+1

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        if (event[iEmt].col() != event[iTop].col())
+          return false;
+      } else {
+        if (event[iEmt].acol() != event[iTop].acol())
+          return false;
+      }
+
+      // Recoiler should now be a W, else something is wrong.
+      if (event[iRec].idAbs() != 24) {
+        cout << " ERROR: recoiler is " << event[iRec].id() << endl;
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:94]:
```suggestion
```
(no "cout" in production code!)

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      double wtT =
+          2. * pRadTop / (pRadEmt * pTopEmt) - pow2(event[iRad].m() / pRadEmt) - pow2(event[iTop].m() / pTopEmt);
+
+      // Histogram weight ratio.
+      wtCorr->fill(wtT / wtW);
+
+      // List relevant properties.
+      if (doList) {
+        cout << "\n now event with sizeOld = " << sizeOld << ", iSys = " << iSys << ", sizeOut = " << sizeOut
+             << scientific << setprecision(3) << ", weight with W = " << wtW << " and with t = " << wtT << endl;
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:119]:
```suggestion
```

File: GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc [hunk 3]
```diff
@@ -20,6 +20,7 @@ using namespace Pythia8;
 
 #include "GeneratorInterface/Pythia8Interface/plugins/ReweightUserHooks.h"
 #include "GeneratorInterface/Pythia8Interface/interface/CustomHook.h"
+#include "GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h"
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:23]:
```suggestion
#include "ReweightUserHooks.h"
#include "GeneratorInterface/Pythia8Interface/interface/CustomHook.h"
#include "TopRecoilHook.h"
```

File: GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc [hunk 4]
```diff
@@ -409,6 +413,13 @@ bool Pythia8Hadronizer::initializeForInternalPartons() {
     fMultiUserHook->addHook(fPowhegHooksBB4L.get());
   }
 
+  bool TopRecoilHook1 = fMasterGen->settings.flag("TopRecoilHook:doTopRecoilIn");
+  if (TopRecoilHook1) {
+    edm::LogInfo("Pythia8Interface") << "Turning on RecoilToTop hook from Pythia8Interface";
+    fTopRecoilHook.reset(new TopRecoilHook());
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:419]: In the master there is a check
```suggestion
    if (!fTopRecoilHook.get())
      fTopRecoilHook.reset(new TopRecoilHook());
```
Why is it not backported here?

File: GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc [hunk 5]
```diff
@@ -568,6 +579,13 @@ bool Pythia8Hadronizer::initializeForExternalPartons() {
...
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:585]: In the master there is a check
```suggestion
    if (!fTopRecoilHook.get())
      fTopRecoilHook.reset(new TopRecoilHook());
```
Why is it not backported here?

**Review by perrotta — COMMENTED** [review]
Why aren't you simply backporting exactly what is in #42180 (+#42313).

**Comment by vslokenb** [comment]
Differences in backport from master (e.g. leaving couts) are not intentional. My apologies!

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
unhold

**Comment by menglu21** [comment]
+1

**Pull request merged**